aip/0194.md

```diff
@@ -45,18 +46,21 @@ The following codes **should not** be retried for any request:
 
 - `OK`: The request succeeded.
 - `CANCELLED`: An application can cancel a request, which **must** be honored.
-- `DEADLINE_EXCEEDED`: An application can set a deadline, which **must** be honored.
-- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never succeed.
-- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be surfaced to the application.
+- `DEADLINE_EXCEEDED`: An application can set a deadline, which **must** be
+  honored.
+- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never
+  succeed.
+- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be
+  surfaced to the application.
 
 ### Generally non-retryable codes:
 
 The following codes generally **should not** be retried for any request:
 
 - `RESOURCE_EXHAUSTED`: This code may be a signal that quota is exhausted.
-  Retries therefore may not be expected to work for several hours; meanwhile the
-  retries may have billing implications. If `RESOURCE_EXHAUSTED` is used for
-  other reasons than quota and the expected time for the resource to become
+  Retries therefore may not be expected to work for several hours; meanwhile
```

Comment:
We thought about that too, but the problem is that `RESOURCE_EXHAUSTED` can mean two different things, and we have no reliable way to know which one.

It can mean (a) being out of a transient quota (e.g. requests per second) or (b) being out of a permanent quota (e.g. disk space). The user experience if we retry (a) is better, but the cost is that the user experience if we retry (b) is downright awful.

The _right_ thing to do here is for us to send and honor a `Retry-After` header, which I believe may be in the works.

Reply:
I would love to see Retry-After become reality!